### PR TITLE
feat: push down window offset to storage

### PR DESCRIPTION
**Pull request by jlapacik**
This PR

* Adds offset parameter to window read cursors
* Updates the planner rule to push down window offset calls
* Has storage expose window offset capability
* Adds unit tests
* Adds launcher tests

### CODE REVIEW

File: storage/reads/datatypes/storage_common.proto [hunk 1]
```diff
@@ -252,5 +252,6 @@ message ReadWindowAggregateRequest {
   TimestampRange range = 2 [(gogoproto.nullable) = false];
   Predicate predicate = 3;
   int64 WindowEvery = 4;
+  int64 Offset = 6;
   repeated Aggregate aggregate = 5;
```
Comment (fchikwekwe) [inline storage/reads/datatypes/storage_common.proto:256]: is there a reason that these are out of order of their enum(?) values? Just curious.

File: storage/flux/reader.go [hunk 2]
```diff
@@ -619,6 +620,7 @@ func (wai *windowAggregateIterator) Do(f func(flux.Table) error) error {
 
 const (
 	CountKind = "count"
+	SumKind   = "sum"
```
Comment (fchikwekwe) [inline storage/flux/reader.go:623]: oh, was sum not available before? Or are you just using it for the first time in this file?

File: query/stdlib/influxdata/influxdb/rules.go [hunk 3]
```diff
@@ -754,7 +747,8 @@ func isPushableWindow(windowSpec *universe.WindowProcedureSpec) bool {
 	// every and period must be equal
 	// every.months must be zero
 	// every.isNegative must be false
-	// offset: must be zero
+	// offset.months must be zero
+	// offset.isNegative must be false
 	// timeColumn: must be "_time"
 	// startColumn: must be "_start"
 	// stopColumn: must be "_stop"
```
Comment (fchikwekwe) [inline query/stdlib/influxdata/influxdb/rules.go:754]: These pushdown rule validators are well organized. Easy to follow.

File: storage/reads/datatypes/storage_common.proto (hunk 1 above)
Comment (jlapacik) [reply storage/reads/datatypes/storage_common.proto:256]: Not really. I was just going off the field ordering of the procedure spec.

File: storage/flux/reader.go (hunk 2 above)
Comment (jlapacik) [reply storage/flux/reader.go:623]: I used it for the first time in a test file, and storage can't import the universe package, so I needed to add it here.

File: storage/reads/modulo.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+		r += modulus
+	}
+	return r
+}
+
+func WindowStart(t, every, offset int64) int64 {
+	mod := Modulo(t, every)
+	off := Modulo(offset, every)
+	beg := t - mod + off
+	if mod < off {
```
Comment (yzhang1991) [inline storage/reads/modulo.go:15]: I stopped here for a moment to think about what is this case about, I guess a comment would be helpful.

File: storage/reads/array_cursor.gen.go.tmpl [hunk 5]
```diff
@@ -350,11 +349,6 @@ func (c *{{.name}}WindowFirstArrayCursor) Next() *cursors.{{.Name}}Array {
 	c.res.Timestamps = c.res.Timestamps[:0]
 	c.res.Values = c.res.Values[:0]
 
-	every := c.every
```
Comment (yzhang1991) [inline storage/reads/array_cursor.gen.go.tmpl:353]: why is this removed?

Comment (jlapacik) [reply storage/reads/array_cursor.gen.go.tmpl:353]: This was dead code so I decided to remove it and be explicit about what window first array cursors expect. They should not be passed `every = 0` or `every = math.MaxInt64`. In both cases a regular first array cursor should be (and is currently) used.

File: storage/reads/modulo.go (hunk 4 above)
Comment (jlapacik) [reply storage/reads/modulo.go:23]: Added doc comments
